docs/manual/src/udl/callback_interfaces.md

```diff
@@ -18,28 +18,56 @@ by the host operating system (e.g. the key chain).
 
 ```rust,no_run
 trait Keychain: Send + Sync + Debug {
-  pub fn get(key: String) -> Option<String>
-  pub fn put(key: String, value: String)
+  pub fn get(key: String) -> Result<Option<String>, KeyChainError>
+  pub fn put(key: String, value: String) -> Result<(), KeyChainError>
 }
 ```
 
+### Why + Send + Sync?
+
 The concrete types that UniFFI generates for callback interfaces implement `Send`, `Sync`, and `Debug`, so it's safe to
 include these as supertraits of your callback interface trait.  This isn't strictly necessary, but it's often useful.  In
 particular, `Sync + Send` is useful when:
   - Storing `Box<dyn CallbackInterfaceTrait>` types inside a type that needs to be `Send + Send` (for example a UniFFI
     interface type)
   - Storing `Box<dyn CallbackInterfaceTrait>` inside a global `Mutex` or `RwLock`
 
-## 2. Define a callback interface in the UDL.
+
+## 2. Setup error handling
+
+All methods of the Rust trait should return a Result.  The error half of that result must
+be an [error type defined in the UDL](./errors.md).
+
+It's also, currently allowed for callback interface methods to return a regular
+value rather than a `Result<>`.  However, this is deprecated and we plan to
+remove support at some point.
+
+### Extra requirements for errors used in callback interfaces
+
+In order to support errors in callback interfaces, UniFFI must be able to
+properly [lift the error](../internals/lifting_and_lowering.md).  This means
+that the if the error is described by an `enum` in the UDL it must be truely
+fieldless --  all variants must be unit variants.
```

Comment:
I don't understand this para - (a) you seem to be saying that if a callback interface throws an error, all variants of that error must be fieldless, but that's not correct is it? I'd expect maybe just the "unexpected" variant should be fieldless? (b) but even that's not true - below you have a callback interface throwing `UnexpectedErrorWithReason` and that takes a string.

ISTM the error variant can be anything at all, but the practical limitation is just that it must be createable given just a `uniffi::UnexpectedUniFFICallbackError`?

(Actually, now I get into the impl, I think I better understand - "truely" is doing a lot of heavy lifting here :) So maybe reword that to better convey that the Rust implementation of the error must not have any fields. Looking at the implementation, it's also not clear what happens if that's not true - should we have a uitest for this scenario?

Reply:
I think you might have figured this out, but I wanted to write it out for posterity.  The issue is using the enum-style definition in the UDL, which doesn't list the fields, when in fact the variants do have fields.  It's now an issue because we have to generate read/lift methods for the scaffolding, so UniFFI needs to know how to construct the variants.

I updated the docs, hopefully it's a bit more clear.  A uitest is a great idea, I put one together and tried to make the error messages as obvious as a could -- although they're still not that obvious.  AFAIK, there's no way to do a static assertion for this and have it print out a nice error message.  If anyone knows one, please share.